### PR TITLE
fail if the given context is not absolute

### LINKED ISSUE #899: Watch mode not working when you have a context in your config.
when I use webpack --watch combine a context config, it's not watch the changes.

### PR DESCRIPTION
**Pull request by timse**
**What kind of change does this PR introduce?**

feature

**Did you add tests for your changes?**

N/A

**If relevant, link to documentation update:**

N/A

**Summary**

fixes #899
blantantly stolen from @mxstbr PR #2698
and updated to the latest version of webpack. 
Hope this gets through before conflicts :)

**Does this PR introduce a breaking change?**

No

### CODE REVIEW

**Comment by sokra** [comment]
Could you instead add this precondition to the schema? i. e. as Regexp.

**Comment by timse** [comment]
Sure. Ill take a look at ajv

File: schemas/ajv.absolutePath.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+"use strict";
+
+const os = require("os");
+
+const getErrorFor = (shouldBeAbsolute, data, schema) => {
+	const message = shouldBeAbsolute ?
+		`The provided value ${JSON.stringify(data)} is not an absolute path!${os.EOL}`
```
Comment (sokra) [inline schemas/ajv.absolutePath.js]: Just use always "\n". We never use os specific EOL, so keep it consistent.

File: schemas/ajv.absolutePath.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+		: `A non absolut path is expected. However the provided value ${JSON.stringify(data)} is an absolute path!${os.EOL}`;
```
Comment (sokra) [inline schemas/ajv.absolutePath.js]: non absolut > non-absolute

File: schemas/ajv.absolutePath.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+		message: message,
+		parentSchema: schema,
+	};
+};
+module.exports = (ajv) => ajv.addKeyword("absolutePath", {
+	errors: true,
+	type: "string",
+	compile(expected, schema) {
+		function callback(data) {
+			const passes = expected === /^(?:[a-zA-Z]:)?(?:\/|\\)/.test(data);
```
Comment (sokra) [inline schemas/ajv.absolutePath.js]: Make sure to disallow `/` in windows paths. Windows paths should be `\` only.

Also disallow a lower-case drive letter. We had problems with it.

**Comment by timse** [comment]
done

**Comment by TheLarkInn** [comment]
From angry Appveyor
```
2) Compiler should compile a single file to deep output:
      Uncaught AssertionError: expected Array [ '/what', '\\what\\the' ] to equal Array [ '/what', '/what\\the' ] (at '1', A has '\\what\\the' and B has '/what\\the')
      + expected - actual
       [
         "/what"
      -  "\\what\\the"
      +  "/what\\the"
       ]
      
      at Assertion.value (C:\projects\webpack\node_modules\should\cjs\should.js:305:19)
      at C:\projects\webpack\test\Compiler.test.js:66:29
```

File: test/Compiler-caching.test.js [hunk 4]
```diff
@@ -14,7 +15,7 @@ describe("Compiler (caching)", function() {
 		new WebpackOptionsDefaulter().process(options);
 		options.entry = entry;
 		options.context = path.join(__dirname, "fixtures");
-		options.output.path = "";
+		options.output.path = __dirname;
```
Comment (sokra) [inline test/Compiler-caching.test.js]: This should be `"/"`.

Set `c.outputFileSystem.join` to `function() { return [].join.call(arguments, "/").replace(/\/+/g, "/"); }` this should only add `/` as separator.

File: test/Compiler-caching.test.js [hunk 5]
```diff
@@ -188,12 +190,11 @@ describe("Compiler (caching)", function() {
 			stats.assets[0].emitted.should.be.exactly(true);
 
 			helper.runAgain((stats, files, iteration) => {
-
 				// Cached the second run
 				stats.assets[0].name.should.be.exactly("bundle.js");
 				stats.assets[0].emitted.should.be.exactly(false);
 
-				files["bundle.js"].should.containEql("This is a");
+				files[bundleFilePath].should.containEql("This is a");
```
Comment (sokra) [inline test/Compiler-caching.test.js]: `/bundle.js`

File: test/Compiler.test.js [hunk 6]
```diff
@@ -1,81 +1,86 @@
...
-		var logs = {
-			mkdirp: [],
-			writeFile: [],
-		};
+function compile(entry, options, callback) {
+	var noOutputPath = !options.output || !options.output.path;
+	new WebpackOptionsDefaulter().process(options);
+	options.entry = entry;
+	options.context = path.join(__dirname, "fixtures");
+	if(noOutputPath) options.output.path = __dirname;
```
Comment (sokra) [inline test/Compiler.test.js]: same

File: test/Compiler.test.js [hunk 7]
```diff
@@ -1,81 +1,86 @@
...
+	beforeEach(function() {
+		mainFilePath = path.join(__dirname, "main.js");
+	});
 
+	it("should compile a single file to deep output", function(done) {
 		compile("./c", {
 			output: {
-				path: 'what',
-				filename: 'the' + sep + 'hell.js',
+				path: __dirname,
```
Comment (sokra) [inline test/Compiler.test.js]: `/what`

File: test/Compiler.test.js [hunk 8]
```diff
@@ -1,81 +1,86 @@
...
+				filename: path.join("the","hell.js"),
```
Comment (sokra) [inline test/Compiler.test.js]: `the/hell.js`

File: test/Compiler.test.js [hunk 9]
```diff
@@ -1,81 +1,86 @@
...
 			}
 		}, function(stats, files) {
 			stats.logs.mkdirp.should.eql([
-				'what',
-				'what' + sep + 'the',
+				__dirname,
+				path.join(__dirname, "the"),
```
Comment (sokra) [inline test/Compiler.test.js]:
```
/what
/what/the
```

File: test/Compiler.test.js [hunk 10]
```diff
@@ -1,81 +1,86 @@
...
 			done();
 		});
 	});
+
 	it("should compile a single file", function(done) {
 		compile("./c", {}, function(stats, files) {
-			files.should.have.property("main.js").have.type("string");
-			Object.keys(files).should.be.eql(["main.js"]);
-			var bundle = files["main.js"];
+			files.should.have.property(mainFilePath).have.type("string");
```
Comment (sokra) [inline test/Compiler.test.js]: `/main.js`

File: test/Validation.test.js [hunk 11]
```diff
@@ -187,6 +188,6 @@ describe("Validation", function() {
 				return;
 			}
 			throw new Error("Validation didn't fail");
-		})
+		});
```
Comment (sokra) [inline test/Validation.test.js]: Add two testcases which tests the error message when a non-/absolute is provided.

File: test/Validation.test.js [hunk 12]
```diff
@@ -174,6 +175,65 @@ describe("Validation", function() {
...
+		name: "non-absolute path",
+		config: {
+			entry: "foo.js",
+			output: {
+				filename: "/bar"
+			}
+		},
+		message: [
+			" - configuration.output.filename A non-absolut path is expected. However the provided value \"/bar\" is an absolute path!",
+			"    ({",
```
Comment (sokra) [inline test/Validation.test.js]: This should not be printed.

Add you addition to lib/WebpackOptionsValidationError.js (formatSchema and formatValidationError)

File: test/Validation.test.js [hunk 13]
```diff
@@ -174,6 +175,31 @@ describe("Validation", function() {
...
+	}, {
+		name: "non-absolute path",
+		config: {
+			entry: "foo.js",
+			output: {
+				filename: "/bar"
+			}
+		},
+		message: [
+			" - configuration.output.filename: A non-absolut path is expected. However the provided value \"/bar\" is an absolute path!",
```
Comment (sokra) [inline test/Validation.test.js]: non-absolut -> relative

**Comment by sokra** [comment]
Thanks

**Comment by timse** [comment]
Thanks for merging :)

**Comment by actuallymentor** [comment]
FYI this is terribly inconvenient for people using __dirname for their paths.

**Comment by timse** [comment]
__dirname is absolute, this is what most of this here is about?

on the other hand certain paths should not be absolute as this makes build bound to the machine they are run on.

can you elaborate where this is a problem for you?

**Comment by actuallymentor** [comment]
> this can potentially make a build bound to the machine they are run on

My webpack is coupled to a custom compiler system. This system has a config.js file with useful paths, system data etc. In there paths are established relative to the config file. E.g. __dirname + /../../templates/thing.pug

Absolute paths are not by definition **fixed** paths. The use of __dirname means the absolute path is still relative to the file where it is called.

By prohibiting absolute paths you are trying to protect pad practices (like tying to a single machine) but you are also reducing the flexibility of webpack.

At the very least this behaviour should have a suppression option for cases where users use absolute (but still dynamic) paths.

**Comment by timse** [comment]
what stops you from making them relative against the your app location?

**Comment by actuallymentor** [comment]
Relative paths in node are relative to the place where node is started. E.g. for /app/location/module/hello.js the ./ is /app if I start node from the /app folder.

By using __dirname inside the hello.js it doesn't matter that I start node from /app or /app/location.

Since webpack is started form one place but I might want to start node from another without breaking by code __dirname buffers against this by allowing references relative to the path of the file as opposed to the path where the process was started.

**Comment by timse** [comment]
I ask again, what stops you from making your path relative against the app location?

the documentation specifies, that you need to provide an absolute path for the ouput, and all you then need to do is specify a relative path from there for your assets. I dont understand where this actually is a problem for you.

see: https://webpack.js.org/concepts/output/#components/sidebar/sidebar.jsx

**Comment by actuallymentor** [comment]
Oof, I am an idiot. My apologies.

I had always specified the full path filename in v1, but at the migration from 1 to 2 only was confronted with the error.

Perhaps it is a good idea to add an explanation to the error along the lines of "if you are trying to specify the path to the output please use ...."?

If so I can PR.

**Comment by timse** [comment]
Have to check how the messaging is implemented, if you can give specific hints im sure better doco is always welcome ;)!

**Pull request merged**